File: exercise/scala/day20/app/reindeer/ReindeerService.scala
package reindeer

import java.util.UUID
import scala.collection.mutable

class ReindeerService {
  private var reindeer: mutable.Seq[Reindeer] = mutable.Seq(
    Reindeer(UUID.fromString("40F9D24D-D3E0-4596-ADC5-B4936FF84B19"), "Petar", ReindeerColor.Black)
  )

  def get(id: UUID): Either[ReindeerErrorCode.Value, Reindeer] =
    reindeer.find(_.id == id).toRight(ReindeerErrorCode.NotFound)

  def create(reindeerToCreate: ReindeerToCreate): Either[ReindeerErrorCode.Value, Reindeer] =
    reindeer.find(_.name == reindeerToCreate.name)
      .map(_ => ReindeerErrorCode.AlreadyExist)
      .toLeft(createAndAddReindeer(reindeerToCreate))

  private def createAndAddReindeer(reindeerToCreate: ReindeerToCreate): Reindeer = {
    val newReindeer = Reindeer(UUID.randomUUID(), reindeerToCreate.name, reindeerToCreate.color)
    reindeer = reindeer :+ newReindeer

    newReindeer
  }
}

case class Reindeer(id: UUID, name: String, color: ReindeerColor.Value)

object ReindeerColor extends Enumeration {
  val White, Black, Purple = Value
}

object ReindeerErrorCode extends Enumeration {
  val NotFound, AlreadyExist = Value
}

case class ReindeerResult(id: UUID, name: String, color: ReindeerColor.Value)

object ReindeerResult {
  def from(reindeer: Reindeer): ReindeerResult =
    ReindeerResult(reindeer.id, reindeer.name, reindeer.color)
}

case class ReindeerToCreate(id: UUID, name: String, color: ReindeerColor.Value)

case class ReindeerToCreateRequest(name: String, color: ReindeerColor.Value) {
  def toDomain: ReindeerToCreate = ReindeerToCreate(UUID.randomUUID(), name, color)
}

File: exercise/kotlin/day20/src/main/kotlin/reindeer/ReindeerService.kt
package reindeer

import arrow.core.Either
import arrow.core.left
import arrow.core.right
import kotlinx.serialization.Contextual
import kotlinx.serialization.Serializable
import java.util.*

object ReindeerService {
    private var reindeer = listOf(
        Reindeer(UUID.fromString("40f9d24d-d3e0-4596-adc5-b4936ff84b19"), "Petar", ReindeerColor.BLACK)
    )

    fun get(id: UUID): Either<ReindeerErrorCode, Reindeer> =
        reindeer.find { it.id == id }?.right() ?: ReindeerErrorCode.NOT_FOUND.left()

    fun create(reindeerToCreate: ReindeerToCreate): Either<ReindeerErrorCode, Reindeer> =
        if (reindeer.any { it.name == reindeerToCreate.name }) {
            ReindeerErrorCode.ALREADY_EXIST.left()
        } else {
            val newReindeer = Reindeer(UUID.randomUUID(), reindeerToCreate.name, reindeerToCreate.color)
            reindeer = reindeer + newReindeer
            newReindeer.right()
        }
}

@Serializable
data class Reindeer(@Contextual val id: UUID, val name: String, val color: ReindeerColor)

@Serializable
enum class ReindeerColor {
    WHITE, BLACK, PURPLE
}

enum class ReindeerErrorCode {
    NOT_FOUND, ALREADY_EXIST
}

@Serializable
data class ReindeerToCreate(@Contextual val id: UUID, val name: String, val color: ReindeerColor)

@Serializable
data class ReindeerToCreateRequest(val name: String, val color: ReindeerColor) {
    fun toDomain() = ReindeerToCreate(UUID.randomUUID(), name, color)
}
